### PR TITLE
Se/add share tresor

### PR DESCRIPTION
**Pull request by sarahESL**
Please give comments. Specially on the tests.

### CODE REVIEW

File: src/services/ZeroKitAdapter.js [hunk 1]
```diff
@@ -148,6 +148,22 @@ class ZeroKitAdapter {
...
+				UserService.resolveUserByEmail(email)
+					.then((grantee) => {
+						console.log(tresorId);
+						return this.zeroKit.then(zeroKit =>
+							zeroKit.shareTresor(tresorId, grantee.zerokit_id)
+								.then(OperationId =>
+									userRoutes
+										.collectivizeTresor(UserService.getUserId(),
+											grantee.id, OperationId)
+										.then(() => tresorId)));
```
Comment (theintz) [inline src/services/ZeroKitAdapter.js]: This looks sketchy, please get in touch with the senior devs to figure out how to turn this into prettier code.

File: src/routes/userRoutes.js [hunk 2]
```diff
@@ -34,6 +34,11 @@ const userRoutes = {
 
 		return hcRequest('POST', `${apiUrl}/tresors`, body);
 	},
+	collectivizeTresor(ownerId, granteeId, OperationId) {
```
Comment (theintz) [inline src/routes/userRoutes.js]: This is not a good name, please find something that is simpler and more intuitive for this function.

File: src/services/UserService.js [hunk 3]
```diff
@@ -76,6 +76,14 @@ class UserService {
 			});
 	}
 
+	resolveUserByEmail(email) {
+		return userRoutes.resolveUserId(email)
+			.then((res) => {
```
Comment (Glumli) [inline src/services/UserService.js]: You can also write:
then(res => res.user)

File: src/services/UserService.js [hunk 4]
```diff
@@ -76,6 +76,14 @@ class UserService {
 			});
 	}
 
+	resolveUserByEmail(email) {
+		return userRoutes.resolveUserId(email)
```
Comment (isalga) [inline src/services/UserService.js]: resolveUserByEmail calls to resolveUserId. This looks strange, maybe `userRoutes.resolveUserId` can be refactored by `resolveUser`. It is called in 3 places, so I think it is a good time to refactor it.

File: src/services/UserService.js [hunk 5]
```diff
@@ -76,6 +76,14 @@ class UserService {
 			});
 	}
 
+	resolveUserByEmail(email) {
```
Comment (Glumli) [inline src/services/UserService.js]: We user alias, except of email

Comment (Glumli) [reply src/services/UserService.js]: And maybe make resolveUser use this method:
return resolveUserByAlias(this.getUserAlias)

File: src/services/ZeroKitAdapter.js (hunk 1 above)
Comment (Glumli) [reply src/services/ZeroKitAdapter.js]: Make this less indented. You can come to me so we can do it in pairprogramming

File: src/HealthCloud.js [hunk 6]
```diff
@@ -29,6 +29,7 @@ class HealthCloud {
 		this.updateFhirRecord = fhirService.updateFhirRecord.bind(fhirService);
 		this.deleteRecord = fhirService.deleteRecord.bind(fhirService);
 		this.logout = zeroKitAdapter.logout.bind(zeroKitAdapter);
+		this.share = zeroKitAdapter.shareTresor.bind(zeroKitAdapter);
```
Comment (Glumli) [inline src/HealthCloud.js]: I would call it shareTresor

File: src/routes/userRoutes.js [hunk 7]
```diff
@@ -35,6 +35,12 @@ const userRoutes = {
 		return hcRequest('POST', `${apiUrl}/tresors`, body);
 	},
 
+	grantTresor(ownerId, granteeId, OperationId) {
```
Comment (Glumli) [inline src/routes/userRoutes.js]: Was this collectivizeTresor earlier?
If so you also have to change the name where it was called

File: test/routes/userRoutesTest.js [hunk 8]
```diff
@@ -76,6 +76,15 @@ describe('userRoutes', () => {
 			done();
 		});
 	});
+	it('collectivize tresor passes', (done) => {
+		requestStub.resolves('pass');
+		userRoutes.collectivizeTresor('ownerId', 'granteeId', 'OperationId').then((res) => {
```
Comment (Harsimran1) [inline test/routes/userRoutesTest.js]: method needs to be fixed in tests too.

File: src/services/UserService.js (hunk 4 above)
Comment (isalga) [inline src/services/UserService.js]: resolveUserByEmail calls to resolveUserId. This looks strange, maybe `userRoutes.resolveUserId` can be refactored by `resolveUser`. It is called in 3 places, so I think it is a good time to refactor it.

File: src/HealthCloud.js (hunk 6 above)
Comment (isalga) [reply src/HealthCloud.js]: I prefer share or shareData. `shareTresor` does not give an idea of what SDK does from a SDK user perspective.

Comment (Glumli) [reply src/HealthCloud.js]: @isalga ?

File: test/routes/userRoutesTest.js [hunk 9]
```diff
@@ -76,6 +76,15 @@ describe('userRoutes', () => {
...
+			expect(res).to.equal('pass');
+			expect(requestStub).to.be.calledOnce;
+			expect(requestStub).to.be.calledWith('POST');
+			done();
+		});
+	});
```
Comment (Harsimran1) [inline test/routes/userRoutesTest.js]: method name needs to be changed in tests too.

File: src/services/UserService.js (hunk 4 above)
Comment (Glumli) [reply src/services/UserService.js]: I just created a ticket to refactor the whole userService. But the first refactoring can be done in this PR.

File: src/services/ZeroKitAdapter.js [hunk 10]
```diff
@@ -148,6 +148,27 @@ class ZeroKitAdapter {
...
+		return Promise.all(
+			[
+				UserService.resolveUserByAlias(alias),
+				this.getTresor(),
+			],
+		)
+			.then((result) => {
+				const grantee = result[0];
+				const tresorId = result[1];
+				console.log(tresorId);
```
Comment (Harsimran1) [inline src/services/ZeroKitAdapter.js]: remove the log statements before pushing.

Comment (sarahESL) [reply src/services/ZeroKitAdapter.js]: Ah! I always forget them! :disappointed:  :+1:

File: src/HealthCloud.js [hunk 11]
```diff
@@ -28,6 +28,7 @@ class HealthCloud {
 		this.downloadFhirRecord = fhirService.downloadFhirRecord.bind(fhirService);
 		this.updateFhirRecord = fhirService.updateFhirRecord.bind(fhirService);
 		this.deleteRecord = fhirService.deleteRecord.bind(fhirService);
+		this.shareData = zeroKitAdapter.shareTresor.bind(zeroKitAdapter);
```
Comment (isalga) [inline src/HealthCloud.js]: Please, remind me to update Terra when this is merged.

File: src/services/ZeroKitAdapter.js [hunk 12]
```diff
@@ -148,6 +148,26 @@ class ZeroKitAdapter {
...
+		return Promise.all(
+			[
+				UserService.resolveUserByAlias(alias),
+				this.getTresor(),
+			],
+		)
+			.then((result) => {
+				const grantee = result[0];
+				const tresorId = result[1];
+				return this.zeroKit.then(zeroKit =>
```
Comment (isalga) [inline src/services/ZeroKitAdapter.js]: Would be better to do some promise chaining. Ask me when you have a moment to take a look

File: src/HealthCloud.js [hunk 13]
```diff
@@ -28,6 +28,7 @@ class HealthCloud {
 		this.downloadFhirRecord = fhirService.downloadFhirRecord.bind(fhirService);
 		this.updateFhirRecord = fhirService.updateFhirRecord.bind(fhirService);
 		this.deleteRecord = fhirService.deleteRecord.bind(fhirService);
+		this.shareData = zeroKitAdapter.grantPermission.bind(zeroKitAdapter);
```
Comment (Glumli) [inline src/HealthCloud.js]: forgot to rename here?

File: test/services/zerokitAdapterTest.js [hunk 14]
```diff
@@ -320,7 +325,39 @@ describe('zerokitAdapter', () => {
...
+			.then((res) => {
+				expect(res).to.equal('fakeOperationId');
+				expect(userRoutesResolveUserIdStub).to.be.calledOnce;
+				expect(zkit_sdk.shareTresor).to.be.calledOnce;
+				expect(userRouteVerifyShareAndGrantPermissionStub).to.be.calledOnce;
+				done();
+			});
+	});
+
+	it('grantPermission fails', (done) => {
```
Comment (Glumli) [inline test/services/zerokitAdapterTest.js]: when does grantPermission fail?

File: src/HealthCloud.js (hunk 13 above)
Comment (isalga) [reply src/HealthCloud.js]: grantPermission? that too cryptic for a SDK user. I think shareData is ok

Comment (Glumli) [reply src/HealthCloud.js]: shareAllData or grantAllData?
For me sharing meens sending the data, but we are not sending it, but when the other user requests it, the user can get them.
grantAllPermissions?

File: test/services/zerokitAdapterTest.js (hunk 14 above)
Comment (sarahESL) [reply test/services/zerokitAdapterTest.js]: "grantPermission fails when zerkit shareTresor fails"...would that be enough?

Comment (Glumli) [reply test/services/zerokitAdapterTest.js]: Si

File: src/HealthCloud.js (hunk 13 above)
Comment (sarahESL) [reply src/HealthCloud.js]: Will keep it as is, and change shareData to grantPermission. Let's decide later for refactoring as we are all not on the same page.

**Pull request merged**